### ISSUE TITLE
Slow Tensor Inference on MPS Compared to NumPy - Possible Upsampling Bottleneck

### ISSUE THREAD

**Issue by theOnlyBoy** [ultralytics/ultralytics#19631]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi team!

Thanks for the amazing YOLOv11 library—truly top-notch.
I’m stuck on an issue for quite a long time, truly.
So... SOS.

I’m running batched PyTorch tensors (`torch.rand(10, 3, 640, 640, device="mps")`) directly into inference on MPS, skipping NumPy preprocessing. Tensor performance lags (~0.07-0.08s per image) versus NumPy (~0.02-0.05s), despite `.contiguous()`, `torch.no_grad()`, warm-up, and manual `non_max_suppression()` batch splitting. Postprocessing is a bottleneck (~0.4s vs. expected ~0.01s), and inference takes ~0.35s. Profiling suggests ~40% in upsampling ops and ~30% in other layers (e.g., conv), hinting at MPS optimization or stride issues.

Am I missing a trick here? I'd like to share more profiling data if it helps—just need a nudge in the right direction.
Cheers!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @theOnlyBoy, thank you for your interest in Ultralytics 🚀! We appreciate your detailed question and your enthusiasm for YOLOv11. 

We recommend checking out our [Docs](https://docs.ultralytics.com/) for guidance, especially the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage sections, which might offer insights into your setup. 

If this is a 🐛 Bug Report, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the issue? This will help us debug the behavior you're experiencing on MPS more effectively.

If this is a ❓ Question, it's great that you've included profiling data and observations! To assist further, we recommend sharing additional details such as:
- Your environment details (macOS version, PyTorch version, MPS details, etc.)
- Code snippets for the inference and postprocessing steps that showcase the issue
- Any additional profiling data that highlights potential bottlenecks

In the meantime, ensure your environment is up-to-date. Upgrade to the latest `ultralytics` package and dependencies in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) by running:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in verified environments that include all dependencies preinstalled (e.g., [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/)):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Join the Community

Engage with the Ultralytics community for more insights and support:
- For real-time discussions, join our [Discord](https://discord.com/invite/ultralytics) 🎧.
- For long-form discussions, visit [Discourse](https://community.ultralytics.com/).
- Connect with the community on our [Subreddit](https://reddit.com/r/Ultralytics) for shared knowledge.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, verifying the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu.

⚡ This is an automated response to ensure your issue is acknowledged. An Ultralytics engineer will review and assist you further soon.

**Comment by Y-T-G** [comment]
ONNX maps Upsampling/Resize Ops to CPU.

https://github.com/microsoft/onnxruntime/issues/14070

Maybe for this reason.

**Comment by Y-T-G** [comment]
For post-processing speed up, you can export the model with `nms=True`. You should in general be using CoreML if you just want inference.

**Comment by theOnlyBoy** [comment]
Unfortunately, using CoreML is not an option. I'd like to understand the issue and fix it.

**Comment by Y-T-G** [comment]
You can export to TorchScript with `nms=True` if you want to stay with PyTorch

**Comment by theOnlyBoy** [comment]
I will try it, thx. But still, if I stay with pt, I'd like to understand the reason why it takes more time with pure tensors than with numpys. Reconverting tensors to numpys doesn't look like an optimized approach.

**Comment by Y-T-G** [comment]
If you're referring to those same layers, then I don't understand what you mean by passing numpy. Because all layers only process tensors, not numpy arrays.

**Comment by theOnlyBoy** [comment]
Maybe my question abt upsampling is incorrect.

The main issue is: using numpy images is faster than using tensors made from those numpys, optimized, resized to 640px, etc. I'd like to figure out why since yolo shouldn't do any preprocessing.

**Comment by Y-T-G** [comment]
Are they the same size?

**Comment by Y-T-G** [comment]
Is the inference slower or the pre and post processing?

**Comment by theOnlyBoy** [comment]
Yup. Same size. Tried different batch sizes 1..N, no difference. Tensors are always slower for some reason.

**Comment by Y-T-G** [comment]
Is it slower if you do:
`model.model(torch.randn(1,3,640,640, device="mps"))`

**Comment by theOnlyBoy** [comment]
@Y-T-G I tried different options, even changed `predictor.py` to make sure there are no extra calculations.
I’ve narrowed it down further with a raw test. Tensor on MPS is still slow:

```
Tensor: ~0.721s total (~0.072s per image)
NumPy (converted to tensor): ~0.242s total (~0.024s per image)
```

Size 640x640 produces the same results.
```
Testing Tensors:
YOLO predict total time (tensor): 0.733414s
Avg time per image (tensor): 0.073341s

Testing NumPy Input (Raw):
YOLO predict total time (numpy raw): 0.193810s
Avg time per image (numpy raw): 0.019381s
```

```python
model = YOLO("yolo11m.pt").to("mps")
batch_tensor = torch.rand(10, 3, 640, 640, device="mps").contiguous()
numpy_batch = [np.random.randint(0, 255, (640, 640, 3), dtype=np.uint8) for _ in range(10)]
tensor_batch = torch.from_numpy(np.stack(numpy_batch)).permute(0, 3, 1, 2).float().to("mps").contiguous() / 255.0

# Tensor
with torch.no_grad():
    preds = model.model(batch_tensor)[0]
results_tensor = non_max_suppression(preds)

# NumPy Raw
with torch.no_grad():
    preds = model.model(tensor_batch)[0]
results_numpy = non_max_suppression(preds)
```

**Comment by Y-T-G** [comment]
I think you should test with actual images instead of random values

**Comment by Y-T-G** [comment]
Is the inference part slower? Without non_max_suppression?

**Comment by theOnlyBoy** [comment]
@Y-T-G  Yup, the ~same difference in numbers:
```
Testing Tensor Input (CPU-to-MPS):
YOLO predict total time (tensor): 0.467458s
Avg time per image (tensor): 0.046746s

Testing NumPy Input (Raw):
YOLO predict total time (numpy raw): 0.138193s
Avg time per image (numpy raw): 0.013819s
```

**Comment by theOnlyBoy** [comment]
Hey guys, I’ve dug deep, and it’s weirder than ever—MPS gives numpy-derived tensors a massive edge over raw tensors, even when we make them identical:
```
Raw Tensor (NumPy Path, Post Warm-Up): ~0.586s (~0.059s per image)
NumPy Tensor (Post Warm-Up): ~0.182s (~0.018s per image)
``
```python
numpy_batch = [np.random.randint(0, 255, (640, 640, 3), dtype=np.uint8) for _ in range(10)]
numpy_array = np.stack(numpy_batch)
numpy_tensor = torch.from_numpy(numpy_array).permute(0, 3, 1, 2).float().to("mps").contiguous() / 255.0
raw_tensor = torch.from_numpy(numpy_array).permute(0, 3, 1, 2).float().to("mps").contiguous() / 255.0  # Same path
# Warm-up both
with torch.no_grad():
    model.model(numpy_tensor)
    model.model(raw_tensor)
# Raw
with torch.no_grad():
    preds = model.model(raw_tensor)[0]
results_raw = non_max_suppression(preds)
# NumPy
with torch.no_grad():
    preds = model.model(numpy_tensor)[0]
results_numpy = non_max_suppression(preds)
```

- Both tensors are (10, 3, 640, 640), contiguous, on MPS, from the same numpy array, yet numpy's ~3x faster post-warm-up.
- Warm-up helps raw tensors (~0.07s → ~0.059s), but not to numpy’s level (~0.018s).
- Reuse matters: first run (raw or numpy) is ~0.07s, second on the same tensor is ~0.018s—except raw tensors never catch up fully.

So, the mystery is: why does `torch.from_numpy()` → MPS inference hit so much better than a raw tensor, even when we recreate it the same way? Is MPS caching something numpy-specific (metadata, kernel hints) that `torch.rand()` or cloning misses, any ideas?

Cheers!

**Comment by Y-T-G** [comment]
How are you measuring the time? Are you using`torch.mps.synchronize()` to synchronize before measuring?

**Comment by theOnlyBoy** [comment]
@Y-T-G , I apologize for the wrong numbers above, not sure how I got them (probably because of a deep night time).
Plot twist—`non_max_suppression()` is what’s killing raw tensor performance on MPS, not inference.

New test with `torch.mps.synchronize()`:

Raw Tensor (Inference Only): ~0.121s (~0.012s per image)
NumPy Tensor (Inference Only): ~0.168s (~0.017s per image)
With NMS (Before): Raw ~0.586s (~0.059s), NumPy ~0.182s (~0.018s)

```python
import torch
import numpy as np
from ultralytics import YOLO
import time

model = YOLO("weights/yolo/yolo11m.pt").to("mps")

numpy_batch = [np.random.randint(0, 255, (640, 640, 3)) for _ in range(10)]
numpy_array = np.stack(numpy_batch)
numpy_tensor = torch.from_numpy(numpy_array).permute(0, 3, 1, 2).float().to("mps").contiguous() / 255.0
raw_tensor = torch.from_numpy(numpy_array).permute(0, 3, 1, 2).float().to("mps").contiguous() / 255.0
with torch.no_grad():
    model.model(numpy_tensor)  # Warm-up
    model.model(raw_tensor)

torch.mps.synchronize()

# Raw
start = time.time()
with torch.no_grad():
    preds = model.model(raw_tensor)[0]
total_raw = time.time() - start  # ~0.012s per image

# NumPy
start = time.time()
with torch.no_grad():
    preds = model.model(numpy_tensor)[0]
total_numpy = time.time() - start  # ~0.017s per image

# With NMS (prior): Raw ~0.059s, NumPy ~0.018s
```

Discovery:

Inference alone: Raw tensors are faster `(~0.012s vs. ~0.017s)`—MPS works fine post-warm-up.
Add `non_max_suppression()`: Raw jumps to `~0.059s (extra ~0.047s)`, numpy stays `~0.018s (extra ~0.001s)`.

NMS chokes on raw tensor preds but flies with numpy-derived preds—same shape, layout, device. Is there an optimization in `non_max_suppression()` tied to numpy tensor origins (metadata, stride)? Any fix to make raw tensors match?

**Comment by Y-T-G** [comment]
`torch.mps.synchronize()` should be called after each inference and even after each `non_max_suppression()` call.

**Comment by theOnlyBoy** [comment]
@Y-T-G Ok, no problem, although it doesn't change anything.
With NMS:
```
Testing Raw Tensor (NumPy Path, Post Warm-Up):
YOLO predict total time (raw tensor): 0.523569s
Avg time per image (raw tensor): 0.052357s

Testing NumPy Tensor (Post Warm-Up):
YOLO predict total time (numpy raw): 0.178426s
Avg time per image (numpy raw): 0.017843s
```
Without NMS:
```
Testing Raw Tensor (NumPy Path, Post Warm-Up):
YOLO predict total time (raw tensor): 0.120746s
Avg time per image (raw tensor): 0.012075s

Testing NumPy Tensor (Post Warm-Up):
YOLO predict total time (numpy raw): 0.168573s
Avg time per image (numpy raw): 0.016857s
```

**Comment by Y-T-G** [comment]
Also what happens if you change the order of inference? numpy first and torch later?

**Comment by theOnlyBoy** [comment]
Figured it out: warming up the model alone doesn’t help much—the first `non_max_suppression()` call on numpys or tensors still slowing things down until pre-called. Thanks for the nudge @Y-T-G.
